### PR TITLE
[Hugging Face Classifier Trainer] Create Hugging Face classifier trainer

### PR DESCRIPTION
**Pull request by davesh0812**
Created the Hugging Face Classifier Trainer Function:

Supporting 2 handlers:

1. train - Training the given classifier class with the relevant params
2. optimize - Optimizing a given HF model and outputs ONNX model.

### CODE REVIEW

File: hugging_face_classifier_trainer/item.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+apiVersion: v1
+categories:
+- machine-learning
+- model-training
+description: Automatic train, evaluate and predict functions for the ML frameworks
+  - Scikit-Learn, XGBoost and LightGBM.
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/item.yaml:6]: ...

File: hugging_face_classifier_trainer/item.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+doc: ''
+example: hugging_face_classifier_trainer.ipynb
+generationDate: 2022-08-28:17-25
+hidden: false
+icon: ''
+labels:
+  author: davids
+maintainers: []
+marketplaceType: ''
+mlrunVersion: 1.3.0-rc29
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/item.yaml:16]: no good to have rc in the hub

File: hugging_face_classifier_trainer/item.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  image: mlrun/ml-models
+  kind: job
+  requirements:
+  - onnx~=1.10.1
+  - onnxruntime~=1.8.1
+  - optimum
+  - transformers
+  - datasets
+url: ''
+version: 1.1.0
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/item.yaml:32]: @yonishelach is this ok?

File: hugging_face_classifier_trainer/item.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  filename: hugging_face_classifier_trainer.py
+  handler: train
+  image: mlrun/ml-models
+  kind: job
+  requirements:
+  - onnx~=1.10.1
+  - onnxruntime~=1.8.1
+  - optimum
+  - transformers
+  - datasets
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/item.yaml:29]: should be frozen like the onnx stuff

File: hugging_face_classifier_trainer/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+mlrun
+transformers~=4.26.0
+numpy~=1.21.6
+pandas~=1.3.5
+datasets
+plotly==5.11.0
+onnx~=1.10.1
+onnxruntime
+optimum
+tensorflow~=2.3
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/requirements.txt]: please freeze all but MLRun

File: hugging_face_classifier_trainer/test_hugging_face_classifier_trainer.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+import mlrun
+import pytest
+from mlrun import import_function
+
+MODELS = [
+    ("sklearn.linear_model.LinearRegression", "regression"),
+    ("sklearn.ensemble.RandomForestClassifier", "classification"),
+    ("xgboost.XGBRegressor", "regression"),
+    ("lightgbm.LGBMClassifier", "classification"),
+]
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/test_hugging_face_classifier_trainer.py:26]: ?

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 7]
```diff
@@ -0,0 +1,794 @@
...
+                context.logger.info(
+                    "not all of the columns to drop in the dataset, drop columns process skipped"
+                )
+    return dataset, label_columns
+
+
+# ---------------------- Hugging Face Trainer --------------------------------
+
+
+def _create_compute_metrics(metrics: List[str]):
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py:544]: please add doc string and return type hint

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 8]
```diff
@@ -0,0 +1,794 @@
...
+                metric
+            ]
+            metric_dict_results[metric] = metric_res
+
+        return metric_dict_results
+
+    return _compute_metrics
+
+
+def _edit_columns(
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py:561]: please add doc string

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 9]
```diff
@@ -0,0 +1,794 @@
...
+    dataset: Dataset,
+    drop_columns: List[str] = None,
+    rename_columns: [str, str] = None,
+):
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py:565]:
```suggestion
) -> Dataset:
```

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 10]
```diff
@@ -0,0 +1,794 @@
...
+    num_of_train_samples: int = None,
+    train_test_split_size: float = None,
+    random_state: int = None,
+):
+    """
+    Loading the dataset and editing the columns
+
+    :param context:                 MLRun contex
+    :param dataset_name:            The name of the dataset to get from the HuggingFace hub
+    :param label_name:
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py:587]: missing doc

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 11]
```diff
@@ -0,0 +1,794 @@
...
+
+def _prepare_dataset(
+    context: MLClientCtx,
+    dataset_name: str,
+    label_name: str = None,
+    drop_columns: Optional[List[str]] = None,
+    num_of_train_samples: int = None,
+    train_test_split_size: float = None,
+    random_state: int = None,
+):
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py:581]:
```suggestion
) -> Tuple[Dataset, Dataset]:
```

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 12]
```diff
@@ -0,0 +1,794 @@
...
+    :param text_col:                The input text column un the dataset.
+    :param num_of_train_samples:    Max number of training samples, for debugging.
+    :param train_test_split_size:   Should be between 0.0 and 1.0 and represent the proportion of the dataset to include
+                                    in the test split.
+    :param metrics:                 List of different metrics for evaluate the model such as f1, accuracy etc.
+    :param random_state:            Random state for train_test_split
+    """
+
+    if not label_name or not pretrained_tokenizer:
+        raise mlrun.errors.MLRunRuntimeError(
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py:667]: I think `MLRunInvalidArgumentError` is more fitting.

Also, if they are mandatory, why they are optional in the func header? (`= None`) Simply make them mandatory and you can remove this validation.

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 13]
```diff
@@ -0,0 +1,794 @@
...
+            "Must provide label_names and pretrained_tokenizer"
+        )
+
+    if train_test_split_size is None:
+        context.logger.info(
+            "test_set or train_test_split_size are not provided, setting train_test_split_size to 0.2"
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py:673]:
```suggestion
            "'test_set' or 'train_test_split_size' are not provided, setting train_test_split_size to 0.2"
```

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 14]
```diff
@@ -0,0 +1,794 @@
...
+        train_dataset, test_dataset = _prepare_dataset(
+            context,
+            hf_dataset,
+            label_name,
+            drop_columns,
+            num_of_train_samples,
+            train_test_split_size,
+            random_state=random_state,
+        )
+    elif dataset:
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py:694]: if both `hf_dataset` and `dataset` are not given? You should catch it with an `else`

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 15]
```diff
@@ -0,0 +1,794 @@
...
+    """
+    Optimizing the transformer model using ONNX optimization.
+
+
+    :param model_path:          The path of the model to optimize.
+    :param model_name:          Name of the optimized model.
+    :param target_dir:          The directory to save the ONNX model.
+    :param optimization_level:  Optimization level performed by ONNX Runtime of the loaded graph. (default is 1)
+    """
+    from optimum.onnxruntime import (ORTModelForSequenceClassification,
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py]:
```suggestion
    # We import these in the function scope so ONNX won't be mandatory for the other handlers:
    from optimum.onnxruntime import (ORTModelForSequenceClassification,
```

**Review by guy1992l — CHANGES_REQUESTED** [review]
The notebook is also pretty messy... A lot of old comments and missing notes, please go over it again more carefully.

File: hugging_face_classifier_trainer/item.yaml (hunk 3 above)
Comment (yonishelach) [reply hugging_face_classifier_trainer/item.yaml:31]: We don't have a convention to function version yet, I think its better to start with `1.0.0` or even with `0.0.1` if its just an initial form of the function (I guess it will be updated when HuggingFace will be in `mlrun/frameworks`)
```suggestion
version: 0.0.1
```

**Review by yonishelach — CHANGES_REQUESTED** [review]
Looks good overall, just a minor comment, every change in the code requires to generate new `function.yaml` so it can be up to date with the code (and the minor version change I mentioned above)

**Review by yonishelach — APPROVED** [review]

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 16]
```diff
@@ -656,14 +680,9 @@ def train(
     :param random_state:            Random state for train_test_split
     """
 
-    if not label_name or not pretrained_tokenizer:
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py]: You removed this yet the parameters are still optional.

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py [hunk 17]
```diff
@@ -692,12 +711,15 @@ def preprocess_function(examples):
...
         )
-
         train_dataset, test_dataset = train_test_split(
             dataset, test_size=train_test_split_size, random_state=random_state
         )
         train_dataset = Dataset.from_pandas(train_dataset)
         test_dataset = Dataset.from_pandas(test_dataset)
+    else:
+        raise mlrun.errors.MLRunInvalidArgumentError(
+            "You have to provide 'hf_dataset' or 'dataset'"
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/hugging_face_classifier_trainer.py:721]: A good exception should print the current errored state -> tell why it was raised -> then, if possible, tell how to overcome it.

So, in this case for example:
```suggestion
            "Training data was not provided. A training dataset is mandatory for training. Please provide a training set using one of the arguments 'hf_dataset' or 'dataset'."
```

File: hugging_face_classifier_trainer/item.yaml [hunk 18]
```diff
@@ -2,8 +2,7 @@ apiVersion: v1
 categories:
 - machine-learning
 - model-training
-description: Automatic train, evaluate and predict functions for the ML frameworks
-  - Scikit-Learn, XGBoost and LightGBM.
+description: Automatic train and optimize functions for Hugging Face frameworks
```
Comment (guy1992l) [inline hugging_face_classifier_trainer/item.yaml:5]:
```suggestion
description: Automatic train and optimize functions for the HuggingFace framework.
```

**Review by guy1992l — CHANGES_REQUESTED** [review]
The notebook is still not looking good. Format the code blocks better and add richer explanations. Go over the text and edit it out, there are a lot of weird text, see "MLhandlers" in the caption where you import the function for example.

**Review by yonishelach — CHANGES_REQUESTED** [review]
Not ready to be merged yet, still need to fix what @guy1992l mentioned above

File: hugging_face_classifier_trainer/hugging_face_classifier_trainer.py (hunk 16 above)
Comment (davesh0812) [reply hugging_face_classifier_trainer/hugging_face_classifier_trainer.py]: I know, there is new error if the tokenizer didn't provided, and about the labels I treat it like hugging face and set the default to be 'labels'

**Review by guy1992l — APPROVED** [review]
Thanks David, great work!
Add a description to the PR and Yoni should merge it :+1:

**Review by yonishelach — APPROVED** [review]
